File: c/stack_kata/src/stack.c
/*
 * stack.c
 *
 *  Created on: Nov 17, 2012
 *      Author: dough
 */

#include "stack.h"

static int elements_in_stack;
static void* elements[MAX_SIZE];
static uint32_t _size;

int Stack_init(const uint32_t size)
{
	int result = STACK_NO_ERR;
	if(size > MAX_SIZE)
	{
		_size = MAX_SIZE;
		result = STACK_ERR_MAX_SIZE_EXCEEDED;
	}
	else
	{
		_size = size;
	}

	elements_in_stack = 0;
	return result;
}

bool Stack_isEmpty(void)
{
	if(elements_in_stack)
	{
		return false;
	}
	else
	{
		return true;
	}
}

int Stack_push(void *element)
{
	int result;
	if(elements_in_stack == _size)
	{
		result = STACK_ERR_OVERFLOW;
	}
	else
	{
		elements[elements_in_stack++] = element;
		result = STACK_NO_ERR;
	}

	return result;
}

void* Stack_pop(void)
{
	if(Stack_isEmpty())
	{
		return STACK_ERR_UNDERFLOW;
	}
	else return elements[--elements_in_stack];
}

void*	Stack_peek(void)
{
	if(Stack_isEmpty())
	{
		return STACK_ERR_UNDERFLOW;
	}
	return elements[elements_in_stack];
}

File: python/stack.py
'''
Created on Nov 11, 2012

@author: Asus
'''

class StackEmptyError(Exception): pass

class Stack(object):
    '''
    My Stack
    '''

    def __init__(self):
        '''
        Constructor
        '''
        self._elements = []
    
    def is_empty(self):
        if len(self._elements) == 0:
            is_empty = True
        else:
            is_empty = False
        return is_empty
    
    def push(self, element):
        self._elements.append(element)
        
    def pop(self):
        if self.is_empty():
            raise StackEmptyError
        else:
            return self._elements.pop(-1)
        
    def peek(self):
        if self.is_empty():
            raise StackEmptyError
        return self._elements[-1]